core/src/calleeFunctions/index.ts

```diff
@@ -20,30 +27,95 @@ const allCalleeFunctions: Record<CalleeNames, CalleeFunctions> = {
     const collateral = getCollateralConfigByType(collateralType);
-    const calleeFuctions = allCalleeFunctions[collateral.exchange.callee];
-    if (!calleeFuctions) {
+    const calleeFuctions = {} as Record<string, CalleeFunctions>;
+    Object.entries(collateral.exchanges).forEach(([key, value]) => {
+        calleeFuctions[key] = allCalleeFunctions[value.callee];
+    });
+    if (!Object.keys(calleeFuctions).length) {
         throw new Error(`Unsupported collateral type "${collateralType}"`);
     }
-    return await calleeFuctions.getCalleeData(network, collateral, profitAddress);
+    const calleesData = {} as Record<string, string>;
+    Object.entries(calleeFuctions).forEach(async ([key, value]) => {
+        calleesData[key] = await value.getCalleeData(network, collateral, profitAddress);
+    });
+    return calleesData;
 };
 
-const _getMarketPrice = async function (
+export const getMarketData = async function (
     network: string,
     collateralSymbol: string,
     amount: BigNumber = new BigNumber('1')
-): Promise<BigNumber> {
+): Promise<Record<string, MarketData>> {
     const collateral = getCollateralConfigBySymbol(collateralSymbol);
-    const calleeFuctions = allCalleeFunctions[collateral.exchange.callee];
-    if (!calleeFuctions) {
+    const calleeIdAndFuctions = {} as Record<string, CalleeFunctions>;
+    Object.keys(collateral.exchanges).forEach(id => {
+        calleeIdAndFuctions[id] = allCalleeFunctions[collateral.exchanges[id].callee];
+    });
+    if (!Object.keys(calleeIdAndFuctions).length) {
         throw new Error(`Unsupported collateral symbol "${collateralSymbol}"`);
     }
-    return await calleeFuctions.getMarketPrice(network, collateral, amount);
+    const marketData = {} as Record<string, MarketData>;
+    const gasPrice = await getGasPriceForUI(network);
+    const exchangeFee = gasPrice.multipliedBy(722651 - 145438);
+    for (const id in calleeIdAndFuctions) {
+        let marketUnitPrice: BigNumber;
+        try {
+            marketUnitPrice = await calleeIdAndFuctions[id].getMarketPrice(network, collateral, amount);
+        } catch {
+            marketUnitPrice = new BigNumber(NaN);
+        }
+        const data = {
+            marketUnitPrice: marketUnitPrice ? marketUnitPrice : new BigNumber(NaN),
+            exchangeFee,
+        } as MarketData;
+        if (collateral.exchanges[id].hasOwnProperty('route')) {
+            data.route = (collateral.exchanges[id] as RegularCalleeConfig).route;
+        } else {
+            data.token0 = (collateral.exchanges[id] as UniswapV2LpTokenCalleeConfig).token0;
+            data.token1 = (collateral.exchanges[id] as UniswapV2LpTokenCalleeConfig).token1;
+        }
+        marketData[id] = data;
+    }
+    return marketData;
+};
+
+export const getBestMarketData = async function (
+    network: string,
+    collateralSymbol: string,
+    amount: BigNumber = new BigNumber('1')
+): Promise<Record<string, string | BigNumber>> {
```

Comment:
This function should accept already calculated `auction.marketData: Record<string, MarketData>` it should NOT refetch it again (which is a lot of requests + it might change between the block)